### ISSUE TITLE
Config: should throw error for duplicate keys

### ISSUE THREAD

**Issue by CuredByLaughter** [SwiftGen/SwiftGen#654]
We recently had an issue that SwiftGen was not updating generated assets. The asset catalog had been updated (new images had been added to the catalog) but the output from SwiftGen was always that the output had not changed and thus no files were to be written.

The issue has since been located! To create a second set out input/output pairs, a developer merely duplicated the `xcassets:` section in the single `swiftgen.yml` configuration file. (Some commits earlier the file had been adjusted to contain multiple `strings:` as well.)

SwiftGen never complained about this - duplicated keys, neither when running the configuration nor when linting the configuration.
SwiftGen happily consumed the poorly formatted configuration file, ignoring all but the last of each keys (e.g. producing output just for the most recently defined `xcassets:`, ignoring the previously defined one).

Running SwiftGen (swiftgen config run --verbose) in talkative mode revealed the details about which files were inputs, which were outputs, as well as showing the commands being run. Instead of the expected number of commands (one for each main section in the configuration file), we were seeing just one for each key.

[yamllint](https://github.com/adrienverge/yamllint) gladly pointed out errors within the .yml configuration file:
 - error - duplication of key "xcassets" in mapping (key-duplicates)
 - error - duplication of key "plist" in mapping (key-duplicates)
 - error - duplication of key "plist" in mapping (key-duplicates)

I would have expected any of the following: 
 - SwiftGen to have refused to process the configuration file given the errors (duplicated keys).
 - SwiftGen to process the file, covering all of the sections specified (even if they had been duplicated and it is an error to duplicate keys in yaml.).
 - SwiftGen to process the file, as was happening, while letting the user know other sections were being ignored.
 - SwiftGen config lint to point out the error.

Our current solution is to merely have a few swiftgen configuration files (such that no main section/key need be repeated).

Configuration file containing the error: _sample_swiftgen.yml_

```
xcassets:
  inputs:
    - Resources/Common.xcassets
    - Resources/Branded.xcassets
  outputs:
    templateName: swift4
    output: Generated/Assets.swift

xcassets:
  inputs:
    - Resources/YetOther.xcassets
  outputs:
    templateName: swift4
    output: Generated/SomeOtherAssets.swift
```

Running `swiftgen config run --config sample_swiftgen.yml` on the above results in only _SomeOtherAssets.swift_ being generated with no explicit indication that _Assets.swift_ is being ignored.

Version data:
_SwiftGen v6.1.0 (Stencil v0.13.1, StencilSwiftKit v2.7.2, SwiftGenKit v6.1.0)_

**Comment by CuredByLaughter** [comment]
For example, Fastlane is happy to point out such a misconfiguration:

> You defined multiple `before_all` blocks in your `Fastfile`. The last one being set will be used.

**Comment by djbe** [comment]
We use Yams to parse these files, funnily enough they had a similar issue: https://github.com/jpsim/Yams/issues/178

Quote:
> That's actually correct according to the YAML spec. You can confirm this by pasting that yaml content in http://www.yamllint.com/

**Comment by CuredByLaughter** [comment]
Thanks for the reply. I looked into the referenced ticket. I also checked what I thought to be a YAML spec: https://yaml.org/spec/1.2/spec.html

There I can find references to JSON indicating that the Keys in JSON SHOULD be unique and that keys in YAML MUST be unique. 

Thus it appears perhaps yams, while reading a file, inserts key-value pairs as it finds them (from top to bottom), overriding any for an existing key, maintaining internally its representation of valid YAML without any non-unique keys. 

So Yams itself does not allow maintaining non-unique keys but at the same time it does not indicate that the input is invalid YAML because it contains non-unique keys.

What do you think @djbe ?

**Comment by CuredByLaughter** [comment]
And here the same thing in [go-yaml](https://github.com/go-yaml/yaml/issues/154)

**Comment by djbe** [comment]
Well looking at the go-yaml issue, it seems the language of the 1.1 spec allows reading duplicate keys (and only throw a warning):
> It is an error for two equal keys to appear in the same mapping node. In such a case the YAML processor may continue, ignoring the second key: value pair and issuing an appropriate warning. This strategy preserves a consistent information model for one-pass and random access applications.

The 1.2 spec is stricter and no longer contains that bit.

We're completely dependent on `Yams` though, so it'll need to be fixed there before we can do anything about it.

**Comment by AliSoftware** [comment]
@PF93mc8y7erq92qTmTjJBysALa About your original goal that led to the issue, be aware that **it _is_ possible to declare multiple entries for a given parser (like `xcassets`) in your config file**.

But the solution to do that is _not_ to repeat the key (as indeed that should be invalid for a dictionary-like structure to have duplicate keys in YAML 1.2), but to instead provide an array of entries (instead of a single entry) for that parser.

Quoting the [`Documentation/ConfigFile.md` documentation](https://github.com/SwiftGen/SwiftGen/blob/develop/Documentation/ConfigFile.md):

> Each key corresponding to a SwiftGen parser (`colors`, `coredata`, `fonts`, `ib`, `json`, `plist`, `strings`, `xcassets`, `yaml`) expects the corresponding value to be:
>
> - Either a dictionary, with the keys described below, if you want to invoke the corresponding SwiftGen parser only once (most common use case)
> - Or an array of those dictionaries, in the less common case where you need to invoke that SwiftGen parser multiple times (for example to use one template with some input files and another template for other input files…)


So basically, the config file extract that you provided in your PR description could be changed to this to make a correct config file that would do the thing you expect:
```yaml
xcassets:
  - inputs:
      - Resources/Common.xcassets
      - Resources/Branded.xcassets
    outputs:
      templateName: swift4
      output: Generated/Assets.swift
  - inputs:
      - Resources/YetOther.xcassets
    outputs:
      templateName: swift4
      output: Generated/SomeOtherAssets.swift
```

---

I agree that this doesn't remove the need for also reporting a nice error in case the user erroneously use duplicate keys in their YAML, but as @djbe said we're dependant on https://github.com/jpsim/Yams/issues/178 for that part 🤷

**Comment by CuredByLaughter** [comment]
Thanks @AliSoftware for the detailed, multifaceted reply where you addressed the original intention, referenced the relevant section of the SwiftGen documentation, rewrote the original example in the form that would work, and still made note that these things alone would not provide the error message @PF93mc8y7erq92qTmTjJBysALa expected.
